File: src/report.rs
//! Rendering of aggregated stats as terminal tables or a Markdown document.

use crate::stats::{BigKey, Stats};
use comfy_table::{Cell, CellAlignment, Table};
use humansize::{format_size, DECIMAL};
use std::fmt::Write as _;

pub struct ReportOptions {
    pub show_memory: bool,
    pub show_types: bool,
    pub show_ttl: bool,
    /// Only print the top N types per namespace (0 = all).
    pub top: usize,
    /// How many biggest keys to list (0 = section omitted).
    pub biggest: usize,
}

fn pct(part: u64, whole: u64) -> String {
    if whole == 0 {
        "0.0%".to_string()
    } else {
        format!("{:.1}%", part as f64 / whole as f64 * 100.0)
    }
}

fn ttl_label(ttl: Option<i64>) -> String {
    match ttl {
        Some(-1) => "persistent".to_string(),
        Some(t) if t >= 0 => format!("{t}s"),
        _ => "-".to_string(),
    }
}

/// Sorted (namespace, key-count, total-bytes) rows, biggest first.
fn namespace_rows(stats: &Stats) -> Vec<(&String, u64, u64)> {
    let mut rows: Vec<(&String, u64, u64)> = stats
        .namespaces
        .keys()
        .map(|ns| (ns, stats.namespace_count(ns), stats.namespace_bytes(ns)))
        .collect();
    rows.sort_by_key(|r| std::cmp::Reverse(r.1));
    rows
}

// ---------------------------------------------------------------------------
// Terminal (comfy-table) rendering
// ---------------------------------------------------------------------------

pub fn render(stats: &Stats, opts: &ReportOptions) {
    let ns_rows = namespace_rows(stats);

    println!("\n=== Namespace summary ===");
    let mut t = Table::new();
    let mut header = vec!["Namespace", "Keys", "%"];
    if opts.show_memory {
        header.push("Memory");
    }
    t.set_header(header);
    for (ns, count, bytes) in &ns_rows {
        let mut row = vec![
            Cell::new(ns),
            Cell::new(count).set_alignment(CellAlignment::Right),
            Cell::new(pct(*count, stats.total_keys)).set_alignment(CellAlignment::Right),
        ];
        if opts.show_memory {
            row.push(Cell::new(format_size(*bytes, DECIMAL)).set_alignment(CellAlignment::Right));
        }
        t.add_row(row);
    }
    println!("{t}");
    println!("Total keys: {}", stats.total_keys);

    for (ns, _, _) in &ns_rows {
        let types = &stats.namespaces[*ns];
        let mut rows: Vec<_> = types.iter().collect();
        rows.sort_by_key(|r| std::cmp::Reverse(r.1.count));
        let total = stats.namespace_count(ns);
        let shown = if opts.top == 0 {
            rows.len()
        } else {
            opts.top.min(rows.len())
        };

        println!("\n--- {} ({} keys, {} types) ---", ns, total, rows.len());
        let mut t = Table::new();
        let mut header = vec!["Type", "Keys", "%"];
        if opts.show_memory {
            header.push("Total mem");
            header.push("Avg mem");
        }
        if opts.show_types {
            header.push("Data types");
        }
        if opts.show_ttl {
            header.push("Persist/Expire");
        }
        t.set_header(header);

        for (name, stat) in rows.iter().take(shown) {
            let mut row = vec![
                Cell::new(name),
                Cell::new(stat.count).set_alignment(CellAlignment::Right),
                Cell::new(pct(stat.count, total)).set_alignment(CellAlignment::Right),
            ];
            if opts.show_memory {
                row.push(
                    Cell::new(format_size(stat.total_bytes, DECIMAL))
                        .set_alignment(CellAlignment::Right),
                );
                row.push(
                    Cell::new(format_size(stat.avg_bytes(), DECIMAL))
                        .set_alignment(CellAlignment::Right),
                );
            }
            if opts.show_types {
                row.push(Cell::new(data_types_str(stat)));
            }
            if opts.show_ttl {
                row.push(Cell::new(format!("{}/{}", stat.persistent, stat.expiring)));
            }
            t.add_row(row);
        }
        println!("{t}");
        if shown < rows.len() {
            let hidden: u64 = rows.iter().skip(shown).map(|(_, s)| s.count).sum();
            println!(
                "  … {} more types hidden ({} keys). Use --top 0 to show all.",
                rows.len() - shown,
                hidden
            );
        }
    }

    render_biggest_terminal(stats, opts);
}

fn render_biggest_terminal(stats: &Stats, opts: &ReportOptions) {
    if opts.biggest == 0 {
        return;
    }
    let big = stats.biggest();
    if big.is_empty() {
        return;
    }
    println!("\n=== Biggest {} keys ===", big.len());
    let mut t = Table::new();
    t.set_header(vec!["#", "Size", "Type", "TTL", "Key"]);
    for (i, b) in big.iter().enumerate() {
        t.add_row(vec![
            Cell::new(i + 1).set_alignment(CellAlignment::Right),
            Cell::new(format_size(b.bytes, DECIMAL)).set_alignment(CellAlignment::Right),
            Cell::new(b.data_type.as_deref().unwrap_or("-")),
            Cell::new(ttl_label(b.ttl)),
            Cell::new(&b.key),
        ]);
    }
    println!("{t}");
}

fn data_types_str(stat: &crate::stats::TypeStat) -> String {
    let mut dts: Vec<_> = stat.data_types.iter().collect();
    dts.sort_by(|a, b| b.1.cmp(a.1));
    dts.iter()
        .map(|(k, v)| format!("{k}:{v}"))
        .collect::<Vec<_>>()
        .join(" ")
}

// ---------------------------------------------------------------------------
// Markdown rendering
// ---------------------------------------------------------------------------

/// Escape the pipe and control characters so a key can't break a Markdown table.
fn md_escape(s: &str) -> String {
    s.replace('|', "\\|").replace(['\n', '\r'], " ")
}

/// `\x01` is a control byte; wrap keys in backticks so it renders literally.
fn md_key(s: &str) -> String {
    format!("`{}`", md_escape(s))
}

pub fn render_markdown(stats: &Stats, opts: &ReportOptions) -> String {
    let mut out = String::new();
    let ns_rows = namespace_rows(stats);

    writeln!(out, "# Redis Insights Report\n").unwrap();
    writeln!(out, "- **Total keys scanned:** {}", stats.total_keys).unwrap();
    writeln!(out, "- **Namespaces:** {}", ns_rows.len()).unwrap();
    if opts.show_memory {
        let total_bytes: u64 = ns_rows.iter().map(|(_, _, b)| *b).sum();
        writeln!(
            out,
            "- **Measured memory:** {}",
            format_size(total_bytes, DECIMAL)
        )
        .unwrap();
    }
    out.push('\n');

    // Namespace summary
    writeln!(out, "## Namespace summary\n").unwrap();
    if opts.show_memory {
        writeln!(out, "| Namespace | Keys | % | Memory |").unwrap();
        writeln!(out, "|---|--:|--:|--:|").unwrap();
        for (ns, count, bytes) in &ns_rows {
            writeln!(
                out,
                "| {} | {} | {} | {} |",
                md_escape(ns),
                count,
                pct(*count, stats.total_keys),
                format_size(*bytes, DECIMAL)
            )
            .unwrap();
        }
    } else {
        writeln!(out, "| Namespace | Keys | % |").unwrap();
        writeln!(out, "|---|--:|--:|").unwrap();
        for (ns, count, _) in &ns_rows {
            writeln!(
                out,
                "| {} | {} | {} |",
                md_escape(ns),
                count,
                pct(*count, stats.total_keys)
            )
            .unwrap();
        }
    }
    out.push('\n');

    // Per-namespace type breakdown
    for (ns, _, _) in &ns_rows {
        let types = &stats.namespaces[*ns];
        let mut rows: Vec<_> = types.iter().collect();
        rows.sort_by_key(|r| std::cmp::Reverse(r.1.count));
        let total = stats.namespace_count(ns);
        let shown = if opts.top == 0 {
            rows.len()
        } else {
            opts.top.min(rows.len())
        };

        writeln!(
            out,
            "## `{}` — {} keys, {} types\n",
            md_escape(ns),
            total,
            rows.len()
        )
        .unwrap();

        let mut head = String::from("| Type | Keys | %");
        let mut sep = String::from("|---|--:|--:");
        if opts.show_memory {
            head.push_str(" | Total mem | Avg mem");
            sep.push_str("|--:|--:");
        }
        if opts.show_types {
            head.push_str(" | Data types");
            sep.push_str("|---");
        }
        if opts.show_ttl {
            head.push_str(" | Persist/Expire");
            sep.push_str("|---");
        }
        head.push_str(" |");
        sep.push('|');
        writeln!(out, "{head}").unwrap();
        writeln!(out, "{sep}").unwrap();

        for (name, stat) in rows.iter().take(shown) {
            write!(
                out,
                "| {} | {} | {}",
                md_key(name),
                stat.count,
                pct(stat.count, total)
            )
            .unwrap();
            if opts.show_memory {
                write!(
                    out,
                    " | {} | {}",
                    format_size(stat.total_bytes, DECIMAL),
                    format_size(stat.avg_bytes(), DECIMAL)
                )
                .unwrap();
            }
            if opts.show_types {
                write!(out, " | {}", md_escape(&data_types_str(stat))).unwrap();
            }
            if opts.show_ttl {
                write!(out, " | {}/{}", stat.persistent, stat.expiring).unwrap();
            }
            writeln!(out, " |").unwrap();
        }
        if shown < rows.len() {
            let hidden: u64 = rows.iter().skip(shown).map(|(_, s)| s.count).sum();
            writeln!(
                out,
                "\n_{} more types hidden ({} keys)._",
                rows.len() - shown,
                hidden
            )
            .unwrap();
        }
        out.push('\n');
    }

    render_biggest_markdown(&mut out, &stats.biggest(), opts);
    out
}

fn render_biggest_markdown(out: &mut String, big: &[BigKey], opts: &ReportOptions) {
    if opts.biggest == 0 || big.is_empty() {
        return;
    }
    writeln!(out, "## Biggest {} keys\n", big.len()).unwrap();
    writeln!(out, "| # | Size | Type | TTL | Key |").unwrap();
    writeln!(out, "|--:|--:|---|---|---|").unwrap();
    for (i, b) in big.iter().enumerate() {
        writeln!(
            out,
            "| {} | {} | {} | {} | {} |",
            i + 1,
            format_size(b.bytes, DECIMAL),
            b.data_type.as_deref().unwrap_or("-"),
            ttl_label(b.ttl),
            md_key(&b.key)
        )
        .unwrap();
    }
    out.push('\n');
}

File: internal/report.go
package internal

import (
	"encoding/json"
	"fmt"
	"sort"
	"strings"

	"github.com/dustin/go-humanize"
)

// ReportOptions configures what a report includes.
type ReportOptions struct {
	ShowMemory bool
	ShowTypes  bool
	ShowTTL    bool
	Top        int // only show top N types per namespace (0 = all)
	Biggest    int // how many biggest keys to list (0 = omit)
}

func pct(part, whole int64) string {
	if whole == 0 {
		return "0.0%"
	}
	return fmt.Sprintf("%.1f%%", float64(part)/float64(whole)*100.0)
}

func ttlLabel(ttl *int64) string {
	if ttl == nil {
		return "-"
	}
	switch *ttl {
	case -1:
		return "persistent"
	case -2:
		return "-"
	}
	if *ttl >= 0 {
		return fmt.Sprintf("%ds", *ttl)
	}
	return "-"
}

type nsRow struct {
	name  string
	count int64
	bytes int64
}

func namespaceRows(stats *Stats) []nsRow {
	var rows []nsRow
	for ns := range stats.Namespaces {
		rows = append(rows, nsRow{
			name:  ns,
			count: stats.NamespaceCount(ns),
			bytes: stats.NamespaceBytes(ns),
		})
	}
	sort.Slice(rows, func(i, j int) bool {
		if rows[i].count == rows[j].count {
			return rows[i].name < rows[j].name
		}
		return rows[i].count > rows[j].count
	})
	return rows
}

// RenderMarkdown renders stats as a Markdown document.
func RenderMarkdown(stats *Stats, opts *ReportOptions) string {
	var out strings.Builder
	nsRows := namespaceRows(stats)

	fmt.Fprintf(&out, "# Redis Insights Report\n\n")
	fmt.Fprintf(&out, "- **Total keys scanned:** %d\n", stats.TotalKeys)
	fmt.Fprintf(&out, "- **Namespaces:** %d\n", len(nsRows))
	if opts.ShowMemory {
		var totalBytes int64
		for _, r := range nsRows {
			totalBytes += r.bytes
		}
		fmt.Fprintf(&out, "- **Measured memory:** %s\n", humanize.Bytes(uint64(totalBytes)))
	}
	out.WriteString("\n")

	// Namespace summary
	fmt.Fprintf(&out, "## Namespace summary\n\n")
	if opts.ShowMemory {
		fmt.Fprintf(&out, "| Namespace | Keys | %% | Memory |\n")
		fmt.Fprintf(&out, "|---|---:|---:|---:|\n")
		for _, r := range nsRows {
			fmt.Fprintf(&out, "| %s | %d | %s | %s |\n",
				mdEscape(r.name), r.count, pct(r.count, stats.TotalKeys), humanize.Bytes(uint64(r.bytes)))
		}
	} else {
		fmt.Fprintf(&out, "| Namespace | Keys | %% |\n")
		fmt.Fprintf(&out, "|---|---:|---:|\n")
		for _, r := range nsRows {
			fmt.Fprintf(&out, "| %s | %d | %s |\n",
				mdEscape(r.name), r.count, pct(r.count, stats.TotalKeys))
		}
	}
	out.WriteString("\n")

	// Per-namespace type breakdown
	for _, nsr := range nsRows {
		types := stats.Namespaces[nsr.name]
		var typeRows []typeRowData
		for name, stat := range types {
			typeRows = append(typeRows, typeRowData{name: name, stat: stat})
		}
		sort.Slice(typeRows, func(i, j int) bool {
			if typeRows[i].stat.Count == typeRows[j].stat.Count {
				return typeRows[i].name < typeRows[j].name
			}
			return typeRows[i].stat.Count > typeRows[j].stat.Count
		})

		total := stats.NamespaceCount(nsr.name)
		shown := len(typeRows)
		if opts.Top > 0 && opts.Top < shown {
			shown = opts.Top
		}

		fmt.Fprintf(&out, "## `%s` — %d keys, %d types\n\n", mdEscape(nsr.name), total, len(typeRows))

		head := "| Type | Keys | %"
		sep := "|---|---:|---:"
		if opts.ShowMemory {
			head += " | Total mem | Avg mem"
			sep += "|---:|---:"
		}
		if opts.ShowTypes {
			head += " | Data types"
			sep += "|---"
		}
		if opts.ShowTTL {
			head += " | Persist/Expire"
			sep += "|---"
		}
		head += " |"
		sep += "|"
		fmt.Fprintf(&out, "%s\n%s\n", head, sep)

		for i := 0; i < shown; i++ {
			tr := typeRows[i]
			fmt.Fprintf(&out, "| `%s` | %d | %s",
				mdEscape(tr.name), tr.stat.Count, pct(tr.stat.Count, total))
			if opts.ShowMemory {
				fmt.Fprintf(&out, " | %s | %s",
					humanize.Bytes(uint64(tr.stat.TotalBytes)),
					humanize.Bytes(uint64(tr.stat.AvgBytes())))
			}
			if opts.ShowTypes {
				fmt.Fprintf(&out, " | %s", mdEscape(dataTypesStr(tr.stat)))
			}
			if opts.ShowTTL {
				fmt.Fprintf(&out, " | %d/%d", tr.stat.Persistent, tr.stat.Expiring)
			}
			fmt.Fprintf(&out, " |\n")
		}

		if shown < len(typeRows) {
			var hidden int64
			for i := shown; i < len(typeRows); i++ {
				hidden += typeRows[i].stat.Count
			}
			fmt.Fprintf(&out, "\n_%d more types hidden (%d keys)._\n", len(typeRows)-shown, hidden)
		}
		out.WriteString("\n")
	}

	// Biggest keys
	if opts.Biggest > 0 {
		big := stats.Biggest()
		if len(big) > 0 {
			fmt.Fprintf(&out, "## Biggest %d keys\n\n", len(big))
			fmt.Fprintf(&out, "| # | Size | Type | TTL | Key |\n")
			fmt.Fprintf(&out, "|---:|---:|---|---|---|\n")
			for i, b := range big {
				dt := "-"
				if b.DataType != nil {
					dt = *b.DataType
				}
				fmt.Fprintf(&out, "| %d | %s | %s | %s | `%s` |\n",
					i+1, humanize.Bytes(uint64(b.Bytes)), dt, ttlLabel(b.TTL), mdEscape(b.Key))
			}
			out.WriteString("\n")
		}
	}

	return out.String()
}

// RenderTable renders stats as a terminal table (using plain text formatting).
func RenderTable(stats *Stats, opts *ReportOptions) string {
	var out strings.Builder
	nsRows := namespaceRows(stats)

	fmt.Fprintf(&out, "\n=== Namespace summary ===\n")

	// Header
	header := fmt.Sprintf("%-20s %10s %7s", "Namespace", "Keys", "%")
	if opts.ShowMemory {
		header += fmt.Sprintf(" %10s", "Memory")
	}
	fmt.Fprintf(&out, "%s\n", header)
	fmt.Fprintf(&out, "%s\n", strings.Repeat("-", len(header)))

	for _, r := range nsRows {
		line := fmt.Sprintf("%-20s %10d %7s", trunc(r.name, 20), r.count, pct(r.count, stats.TotalKeys))
		if opts.ShowMemory {
			line += fmt.Sprintf(" %10s", humanize.Bytes(uint64(r.bytes)))
		}
		fmt.Fprintf(&out, "%s\n", line)
	}
	fmt.Fprintf(&out, "Total keys: %d\n", stats.TotalKeys)

	for _, nsr := range nsRows {
		types := stats.Namespaces[nsr.name]
		var typeRows []typeRowData
		for name, stat := range types {
			typeRows = append(typeRows, typeRowData{name: name, stat: stat})
		}
		sort.Slice(typeRows, func(i, j int) bool {
			if typeRows[i].stat.Count == typeRows[j].stat.Count {
				return typeRows[i].name < typeRows[j].name
			}
			return typeRows[i].stat.Count > typeRows[j].stat.Count
		})

		total := stats.NamespaceCount(nsr.name)
		shown := len(typeRows)
		if opts.Top > 0 && opts.Top < shown {
			shown = opts.Top
		}

		fmt.Fprintf(&out, "\n--- %s (%d keys, %d types) ---\n", nsr.name, total, len(typeRows))

		for i := 0; i < shown; i++ {
			tr := typeRows[i]
			line := fmt.Sprintf("  %-40s %8d %7s", trunc(tr.name, 40), tr.stat.Count, pct(tr.stat.Count, total))
			if opts.ShowMemory {
				line += fmt.Sprintf(" %10s %10s",
					humanize.Bytes(uint64(tr.stat.TotalBytes)),
					humanize.Bytes(uint64(tr.stat.AvgBytes())))
			}
			if opts.ShowTypes {
				line += fmt.Sprintf(" %s", trunc(dataTypesStr(tr.stat), 20))
			}
			if opts.ShowTTL {
				line += fmt.Sprintf(" %d/%d", tr.stat.Persistent, tr.stat.Expiring)
			}
			fmt.Fprintf(&out, "%s\n", line)
		}

		if shown < len(typeRows) {
			var hidden int64
			for i := shown; i < len(typeRows); i++ {
				hidden += typeRows[i].stat.Count
			}
			fmt.Fprintf(&out, "  … %d more types hidden (%d keys).\n", len(typeRows)-shown, hidden)
		}
	}

	// Biggest keys
	if opts.Biggest > 0 {
		big := stats.Biggest()
		if len(big) > 0 {
			fmt.Fprintf(&out, "\n=== Biggest %d keys ===\n", len(big))
			fmt.Fprintf(&out, "%-5s %10s %-8s %-12s %s\n", "#", "Size", "Type", "TTL", "Key")
			fmt.Fprintf(&out, "%s\n", strings.Repeat("-", 70))
			for i, b := range big {
				dt := "-"
				if b.DataType != nil {
					dt = *b.DataType
				}
				fmt.Fprintf(&out, "%-5d %10s %-8s %-12s %s\n",
					i+1, humanize.Bytes(uint64(b.Bytes)), dt, ttlLabel(b.TTL), trunc(b.Key, 40))
			}
		}
	}

	return out.String()
}

// RenderJSON renders stats as JSON.
func RenderJSON(stats *Stats) string {
	type jsonReport struct {
		*Stats
		Biggest []BigKey `json:"biggest"`
	}
	b, _ := json.MarshalIndent(jsonReport{Stats: stats, Biggest: stats.Biggest()}, "", "  ")
	return string(b)
}

type typeRowData struct {
	name string
	stat *TypeStat
}

func dataTypesStr(stat *TypeStat) string {
	type dtPair struct {
		name  string
		count int64
	}
	var pairs []dtPair
	for k, v := range stat.DataTypes {
		pairs = append(pairs, dtPair{k, v})
	}
	sort.Slice(pairs, func(i, j int) bool {
		if pairs[i].count == pairs[j].count {
			return pairs[i].name < pairs[j].name
		}
		return pairs[i].count > pairs[j].count
	})
	var parts []string
	for _, p := range pairs {
		parts = append(parts, fmt.Sprintf("%s:%d", p.name, p.count))
	}
	return strings.Join(parts, " ")
}

func mdEscape(s string) string {
	s = strings.ReplaceAll(s, "|", "\\|")
	s = strings.ReplaceAll(s, "\n", " ")
	s = strings.ReplaceAll(s, "\r", " ")
	return s
}

func trunc(s string, max int) string {
	runes := []rune(s)
	if len(runes) <= max {
		return s
	}
	return string(runes[:max-1]) + "…"
}

// PersistRow represents a type with persistent (no-TTL) keys.
type PersistRow struct {
	Namespace string
	KeyType   string
	Keys      int64
	Bytes     int64
}

// PersistentRows returns all (namespace, type) buckets with persistent keys,
// sorted by memory (descending).
func PersistentRows(stats *Stats) []PersistRow {
	var rows []PersistRow
	for ns, types := range stats.Namespaces {
		for t, s := range types {
			if s.Persistent > 0 {
				rows = append(rows, PersistRow{
					Namespace: ns,
					KeyType:   t,
					Keys:      s.Persistent,
					Bytes:     s.PersistentBytes,
				})
			}
		}
	}
	sort.Slice(rows, func(i, j int) bool {
		if rows[i].Bytes != rows[j].Bytes {
			return rows[i].Bytes > rows[j].Bytes
		}
		if rows[i].Keys != rows[j].Keys {
			return rows[i].Keys > rows[j].Keys
		}
		if rows[i].Namespace != rows[j].Namespace {
			return rows[i].Namespace < rows[j].Namespace
		}
		return rows[i].KeyType < rows[j].KeyType
	})
	return rows
}
